pkg/apis/wksprovider/controller/wksctl/machine_actuator.go

```diff
@@ -406,41 +423,144 @@ func (a *MachineActuator) update(ctx context.Context, cluster *clusterv1.Cluster
+				if err := drain.Drain(node, a.clientSet, drain.Params{
+					Force:               true,
+					DeleteLocalData:     true,
+					IgnoreAllDaemonSets: true,
+				}); err != nil {
+					return err
+				}
+			} else if isOriginal {
+				return a.kubeadmUpOrDowngrade(machine, node, installer, version, planKey, planJSON, originalMaster)
+			} else {
+				return a.kubeadmUpOrDowngrade(machine, node, installer, version, planKey, planJSON, secondaryMaster)
 			}
 		}
+		return a.kubeadmUpOrDowngrade(machine, node, installer, version, planKey, planJSON, worker)
+	}
+	if err = a.performActualUpdate(installer, machine, node, nodePlan, c); err != nil {
+		return err
+	}
+	if err = a.setNodeAnnotation(node, planKey, planJSON); err != nil {
+		return err
+	}
+	a.recordEvent(machine, corev1.EventTypeNormal, "Update", "updated machine %s", machine.Name)
+	return nil
+}
+
+func (a *MachineActuator) kubeadmUpOrDowngrade(machine *clusterv1.Machine, node *corev1.Node, installer *os.OS,
+	version, planKey, planJSON string, ntype nodeType) error {
+	b := plan.NewBuilder()
+	b.AddResource(
+		"upgrade:node-unlock-kubernetes",
+		&resource.Run{Script: object.String("yum versionlock delete 'kube*' || true")})
+	b.AddResource(
+		"upgrade:node-install-kubeadm",
+		&resource.RPM{Name: "kubeadm", Version: version, DisableExcludes: "kubernetes"},
+		plan.DependOn("upgrade:node-unlock-kubernetes"))
+	switch ntype {
+	case originalMaster:
+		b.AddResource(
+			"upgrade:node-kubeadm-upgrade",
+			&resource.Run{Script: object.String(fmt.Sprintf("kubeadm upgrade plan && kubeadm upgrade apply -y %s", version))},
+			plan.DependOn("upgrade:node-install-kubeadm"))
+	case secondaryMaster:
+		b.AddResource(
+			"upgrade:node-kubeadm-upgrade",
+			&resource.Run{Script: object.String("kubeadm upgrade node experimental-control-plane")},
+			plan.DependOn("upgrade:node-install-kubeadm"))
+	case worker:
+		b.AddResource(
+			"upgrade:node-kubeadm-upgrade",
+			&resource.Run{Script: object.String(fmt.Sprintf("kubeadm upgrade node config --kubelet-version %s", version))},
+			plan.DependOn("upgrade:node-install-kubeadm"))
+	}
+	b.AddResource(
+		"upgrade:node-kubelet",
+		&resource.RPM{Name: "kubelet", Version: version, DisableExcludes: "kubernetes"},
+		plan.DependOn("upgrade:node-kubeadm-upgrade"))
+	b.AddResource(
+		"upgrade:node-restart-kubelet",
+		&resource.Run{Script: object.String("systemctl restart kubelet")},
+		plan.DependOn("upgrade:node-kubelet"))
+	b.AddResource(
+		"upgrade:node-kubectl",
+		&resource.RPM{Name: "kubectl", Version: version, DisableExcludes: "kubernetes"},
+		plan.DependOn("upgrade:node-restart-kubelet"))
+	b.AddResource(
+		"upgrade:node-lock-kubernetes",
+		&resource.Run{Script: object.String("yum versionlock add 'kube*' || true")},
+		plan.DependOn("upgrade:node-kubectl"))
+
+	p, err := b.Plan()
+	if err != nil {
 		return err
 	}
+	err = installer.SetupNode(&p)
+	if err != nil {
+		log.Infof("Failed to upgrade master node %s: %v", node.Name, err)
```

Comment:
I suspect I might have missed something but this logic appears to upgrade worker node in addition to master notes.  This comment might need to be update.  Same comment for link 546